velox/functions/lib/Re2Functions.cpp

```diff
@@ -383,13 +383,220 @@ class Re2SearchAndExtract final : public VectorFunction {
   const bool emptyNoMatch_;
 };
 
+namespace {
+
+// Sub-pattern's stats in a top-level pattern.
+struct SubPatternStats {
+  SubPatternKind kind;
+  // Count of this patterns in total.
+  size_t count = 0;
+  // First index of this pattern kind.
+  int32_t firstIndex = -1;
+  // Last index of this pattern kind.
+  int32_t lastIndex = -1;
+
+  void update(size_t index) {
+    count++;
+    if (firstIndex == -1) {
+      firstIndex = index;
+    }
+
+    lastIndex = index;
+  }
+};
+
+// Construct SubPatterns from subPatternKinds, subPatternRanges.
+// Caller need to make sure the specified range only contains
+// fixed(kLiteralString, kSingleWildcard) patterns.
+size_t buildFixedSubPatterns(
+    const std::vector<SubPatternKind>& subPatternKinds,
+    const std::vector<std::pair<size_t, size_t>>& subPatternRanges,
+    size_t start,
+    size_t end,
+    std::vector<SubPatternMetadata>& subPatterns) {
+  size_t fixedLength = 0;
+  for (auto i = start; i < end; i++) {
+    if (subPatternKinds[i] == SubPatternKind::kLiteralString) {
+      subPatterns.push_back(
+          {SubPatternKind::kLiteralString,
+           fixedLength,
+           subPatternRanges[i].second});
+    } else if (subPatternKinds[i] == SubPatternKind::kSingleCharWildcard) {
+      subPatterns.push_back(
+          {SubPatternKind::kSingleCharWildcard,
+           fixedLength,
+           subPatternRanges[i].second});
+    } else {
+      VELOX_UNREACHABLE();
+    }
+    fixedLength += subPatternRanges[i].second;
+  }
+
+  return fixedLength;
+}
+
+// Return the length of the fixed part(literal chars or single char wildcard) of
+// the pattern, it is mainly used to get the length of the fixed part in a
+// pattern, so we can extract the fixed part out.
+size_t fixedLength(
+    const std::vector<SubPatternKind>& subPatternKinds,
+    const std::vector<std::pair<size_t, size_t>>& subPatternRanges) {
+  size_t result = 0;
+  for (auto i = 0; i < subPatternKinds.size(); i++) {
+    if (subPatternKinds[i] != SubPatternKind::kAnyCharsWildcard) {
+      result += subPatternRanges[i].second;
+    }
+  }
+
+  return result;
+}
+
+// Get the num of bytes of current char(might be unicode char).
```

Comment:
Return the number of bytes in the specified unicode character. Returns 1 if specified character is not a valid UTF-8.